### ISSUE TITLE
关于 PPO 训练代码的一些讨论

### ISSUE THREAD

**Issue by annn521** [jingyaogong/minimind#564]
Hi! 作者大大您好，最近一直在学习大模型中的强化学习，关于 PPO 训练代码中的一些细节有些疑问。

首先是在计算 advantage 时采用的是句子级细粒度而不是 token 级细粒度，因此也就无法考虑未来收益以及无法使用GAE计算回报 return，这也使得 critic model 的损失计算改变；

<img width="353" height="24" alt="Image" src="https://github.com/user-attachments/assets/3bb59e53-21d1-434e-93ae-3e622125989c" />

其次是 actor 与 ref 的 KL 散度部分，通常是用于添加到 reward 中，但在您的实现中是直接加到了损失中 （我猜想是因为前面采用的句子级的细粒度，因此整个的计算被简化了）；

<img width="271" height="20" alt="Image" src="https://github.com/user-attachments/assets/30bcf8f5-18ad-4a2e-98ec-813439353e6a" />

<img width="476" height="40" alt="Image" src="https://github.com/user-attachments/assets/ad4d4ab5-32ce-4d08-a9a6-2370ebd6d969" />

最后就是对于单个 batch 的数据通常会进行 ppo_epoch 次迭代来加强单个 batch 数据的使用，而在实现中只利用了一次。

因此想要请问一下作者大大上述的改变是出于对训练过程的简化或者是为了适配 minimind 还是其他的目的，感谢您的时间。

**Comment by jingyaogong** [comment]
你的理解没有问题。
这里的“PPO”骨架上是基于 GRPO 写的

* 句级 Advantage
* 不用 GAE
* KL 加到了 loss
* 单次更新

主要区别是保留了 Critic Model（虽然只取最后一个 token 的输出）来做 baseline，GRPO 用组内均值代替。
可以理解为以 GRPO 为基础加 Critic 的简化版 PPO。

| 维度 | 这里的 “PPO” | 标准 GRPO | 标准 PPO |
|------|-------------|-----------|---------|
| **Advantage** | `R - V_last` 句级 `[B]` | `(R-μ)/σ` 句级 `[B*G]` | GAE Token级 `[B,R]` |
| **GAE** | 无 | 无 | 有 |
| **KL 位置** | `loss += β*KL` | `loss_t -= β*KL_t` | `r_t -= β*KL_t` |
| **PPO Epochs** | 1 | 1 | 1~K |

**Comment by annn521** [comment]
感谢作者回复。

**Comment by annn521** [comment]
此外还有一些问题想讨论，不久前 Qwen 以及 Kimi 均推出了他们的 “下一代” 模型 Qwen3-Next 以及 Kimi Linear，都是由线性注意力与全注意力（比例3：1）组成的混合架构。这样的混合架构能够降低训练与推理成本，我想在某种程度上也符合 Minimind （低成本快速搭建大模型），Kimi Linear 在技术报告中也表明他们的混合架构优于了全注意力架构。

想要请问一下作者关于混合架构的一些看法，在类似 Minimind 的低参数量上是否同样可能会有更好的表现，以及您是否尝试过或者计划尝试推出类似的混合架构的 Minimind。感谢您的时间。
